HW2/P5/P5.txt

```diff
@@ -0,0 +1,118 @@
+Author: Xingchi Dai
+Harvard University
+CS205
+HW2
+Problem 5: Simulation and domain decomposition
+
+Part1: Performance results
+
+Performance:
+
+1 thread:
+5.30881548837 simulation frames per second
+5.40018050797 simulation frames per second
+5.41002962789 simulation frames per second
+5.25063781738 simulation frames per second
+5.39368750257 simulation frames per second
+5.46699252481 simulation frames per second
+
+4 threads:
+
+10.6414713292 simulation frames per second
+11.1193694709 simulation frames per second
+11.8052638541 simulation frames per second
+10.9182311353 simulation frames per second
+10.7273122537 simulation frames per second
+11.4661753208 simulation frames per second
+11.8020749211 simulation frames per second
+
+Note: parallel computing shows a good improvement on performance, almost two times better.
+
+
+
+Part3: Spatial decomposition
+Performance data:
+
+1 thread:
+
+420.02683811 simulation frames per second
+435.250297479 simulation frames per second
+441.946079602 simulation frames per second
+446.357670396 simulation frames per second
+439.835851259 simulation frames per second
+409.634288584 simulation frames per second
+417.044310765 simulation frames per second
+
+
+4 threads:
+1208.35093982 simulation frames per second
+1203.76645177 simulation frames per second
+1215.83484046 simulation frames per second
+1221.72583957 simulation frames per second
+1215.42236389 simulation frames per second
+1210.1877142 simulation frames per second
+
+Note: Compared to previous version, we get significantly improved on performance. The reason is we convert our
+algorithm from O(N2) to O(N). The complexity of our algorithm has decreased a lot.
+
+Part3 and 4: Spatially Coherent Sorting/Locking
+Performance data:
+
+Without locks:
+1 thread:
+
+600.858060863 simulation frames per second
+532.272081218 simulation frames per second
+562.544544961 simulation frames per second
+600.771996835 simulation frames per second
+605.484144606 simulation frames per second
+539.789182782 simulation frames per second
+599.794240421 simulation frames per second
+
+4 threads:
+1409.2252175 simulation frames per second
+1408.00731387 simulation frames per second
+1426.86656697 simulation frames per second
+1394.86643034 simulation frames per second
+1388.44132666 simulation frames per second
+1397.38323362 simulation frames per second
+1415.15194883 simulation frames per second
+1435.90348694 simulation frames per second
+1393.56483504 simulation frames per second
+1398.60638564 simulation frames per second
+
+
+
+Note: Since now the point close in space are actually close in memory, we get a little bit higher performance both
```

Comment:
Notice that the parallel gains are (in terms of percentage gain) greater than the serial ones. Why do you think this might be?